README
PerlActor version 0.01
======================

PerlActor is a simple automated executable acceptance test framework
for Perl.  It can be used by Extreme Programming (XP) teams to rapidly
develop executable acceptance tests for their Perl code.

XP mandates that the Customer writes acceptance tests for the application
under development.  The acceptance tests, also known as customer tests,
demonstrate that application features are complete and that they work as
expected.  Ideally, the tests should be directly executable and
automated so that they can be run continuously, without manual effort.

PerlActor enables an XP Customer to write *executable* acceptance
tests as plain text files, using a very simple syntax. Each test consists
of a number of parameterized commands (one per line), which PerlActor
uses to exercise the application. PerlActor parses the test, invokes the
commands with any parameters, and reports the result. The developers provide
glue code to implement the commands required by the tests.

PerlActor allows tests to be grouped into suites so that related tests can be
run as a group.  The Customer can also just place test scripts in a directory
structure and have PerlActor find and execute them all.

The PerlActor approach to acceptance testing has a number of advantages:

1) As the application grows the team will gradually develop a
   comprehensive set of Commands for testing the application.

2) The Customer, independent of the development team, can write and execute
   new tests for the application at any time, using any of the existing Commands.

3) Existing tests may be changed by the Customer at any time, again
   without developer help.


EXAMPLE
=======

The following sample test script is adapted from one of the example
scripts contained in the 'examples/calculator' directory of this
distribution.  It tests a toy calculator "application":

    # Script to check addition
    # 10 + 32 = 42

    # Create a new calculator application
    NewCalculator

    # Ensure that the display reads '0' intially
    CheckDisplayReads  0

    # Press key '1', then key '0'
    PressKeys          1  0
    CheckDisplayReads  10

    PressKeys          +
    CheckDisplayReads  10

    PressKeys          3  2
    CheckDisplayReads  32

    PressKeys          =
    CheckDisplayReads  42


The following is sample output from the included test runner:

    Running Acceptance Tests at Sun Apr 10 17:59:11 GMT 2005
    ==========================================================
    .....
    0.545077 wallclock secs ( 0.42 usr +  0.11 sys =  0.53 CPU)
    Run: 5, Passed: 5, Failed: 0, Aborted: 0.

and with test failure:

    Running Acceptance Tests at Sun Apr 10 18:04:36 GMT 2005
    ==========================================================
    .F...
    1.24811 wallclock secs ( 0.47 usr +  0.07 sys =  0.54 CPU)
    Run: 5, Passed: 4, Failed: 1, Aborted: 0.

    !!!FAILED!!!

    1) FAILED: 'Display is wrong: expected 41, got 42' in scripts/test_addition.pact
    at 'CheckDisplayReads  41', line 37


CREDITS

PerlActor is inspired by a similar open source framework for Java called 
'exactor' published by eXoftware [see http://exactor.sourceforge.net/].

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Test::Unit - needed for running PerlActor's own unit test suite.

COPYRIGHT AND LICENCE

Copyright (C) 2005 by Tony Byrne

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension PerlActor.

0.01  Wed Apr  6 19:59:24 2005
	- original version; created by h2xs 1.23 with options
		-AX -n PerlActor



run_tests.pl
#!/usr/bin/perl -w

use strict;
use lib qw( lib t/tlib );
use Test::Unit::TestRunner;

my @tests;

if (@ARGV)
{
	@tests = @ARGV;
}
else
{
	@tests = ('PerlActor::test::AllTests');
}

print "\nRunning Test at " . `date`;
print "==========================================================\n";
my $testrunner = Test::Unit::TestRunner->new();

while (my $test = shift @tests)
{
	# Turn possible path into package name
	#$test =~ s/^.*test\///;
	$test =~ s/\.pm$//g;
	$test =~ s/\//::/g;

	$testrunner->start($test);
}


t/tlib/PerlActor/Command/Dummy.pm
package PerlActor::Command::Dummy;
use strict;

use base 'PerlActor::Command';

#===============================================================================================
# Public Methods
#===============================================================================================

sub execute
{
	my $self = shift;
	$self->{context}->{dummy_was_here} = 1;
}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

# Keep Perl happy.
1;


examples/calculator/test/PressKeys.pm
package PressKeys;
use strict;

use base 'PerlActor::Command';

#===============================================================================================
# Public Methods
#===============================================================================================

sub execute
{
	my $self = shift;
	my @keys = $self->getParams();
	foreach my $key (@keys)
	{
		$self->{context}->{calculator}->pressKey($key);
	}
}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

# Keep Perl happy.
1;


lib/PerlActor/Object.pm
package PerlActor::Object;
use fields qw( listener context );
use strict;

#===============================================================================================
# Public Methods
#===============================================================================================

sub new
{
	my $proto = shift;
	my $class = ref $proto || $proto;
	my $self = fields::new($class);
	return $self;
}

sub setListener
{
	my $self = shift;
	$self->{listener} = shift;
}

sub setContext
{
	my $self = shift;
	$self->{context} = shift;
}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

# Keep Perl happy.
1;


lib/PerlActor/Script.pm
package PerlActor::Script;
use strict;
use base 'PerlActor::Object';
use fields qw( lines name currentLine currentLineNumber );
use Error qw( :try );

use PerlActor::CommandFactory;
use PerlActor::LineTokenizer;
use PerlActor::Exception::AssertionFailure;
use PerlActor::Exception::CommandFailed; 
use PerlActor::Exception::CommandAborted; 

#===============================================================================================
# Public Methods
#===============================================================================================

sub new
{
	my $proto = shift;
	my $self = $proto->SUPER::new(@_);
	$self->{name} = shift;
	$self->{lines} = [];
	$self->{context} = {};
	return $self;
}

sub setLines
{
	my $self = shift;
	my @lines = @_;
	$self->{lines} = \@lines;
}

sub getName
{
	my $self = shift;
	return $self->{name};
}

sub execute
{
	my $self = shift;
	$self->{listener}->scriptStarted($self);	

	$self->{currentLineNumber} = 0;
	try
	{
		foreach my $line (@{$self->{lines}})
		{
			chomp $line;
			$self->{currentLine} = $line;
			$self->{currentLineNumber}++;			
			my $command = $self->_parseLine($line);
			next unless $command;
			$self->executeCommand($command);
		}
		$self->{listener}->scriptPassed($self);		
		$self->{listener}->scriptEnded($self);
	}
	catch PerlActor::Exception::CommandFailed with
	{
		my $exception = shift;
		$self->{listener}->scriptFailed($self, $exception);
		$self->{listener}->scriptEnded($self);
	}
	catch PerlActor::Exception::CommandAborted with
	{
		my $exception = shift;
		$self->{listener}->scriptAborted($self, $exception);		
	};
}

sub executeCommand
{
	my $self = shift;
	my $command = shift;
	$command->setContext($self->{context});
	$command->setListener($self->{listener});
	$self->{listener}->commandStarted($self, $command);
	my $success;
	try
	{
		$command->execute();
		$self->{listener}->commandPassed($self, $command);				
		$self->{listener}->commandEnded($self, $command);	
	}
	catch PerlActor::Exception::AssertionFailure with
	{
		my $exception = shift;
		$self->{listener}->commandFailed($self, $command, $exception);
		$self->{listener}->commandEnded($self, $command);
		throw PerlActor::Exception::CommandFailed("$exception\n");
	}
	otherwise
	{
		my $exception = shift;
		$self->{listener}->commandAborted($self, $command, $exception);
		throw PerlActor::Exception::CommandAborted("$exception\n");
	};
}

sub parse
{
	my $self = shift;
	return $self->_parseLines(@{$self->{lines}});
}

sub getTraceInfo
{
	my $self = shift;
	my $name = $self->getName();
	return "in $name at '$self->{currentLine}', line $self->{currentLineNumber}";
}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

sub _parseLines
{
	my $self = shift;
	my @lines = @_;
	my @commands;
	foreach my $line (@lines)
	{
		push @commands, $self->_parseLine($line);
	}
	return @commands;
}

sub _parseLine
{
	my ($self, $line) = @_;
	
	my $tokenizer = new PerlActor::LineTokenizer();
	my @tokens = $tokenizer->getTokens($line);

	return unless @tokens;
		
	my $factory = new PerlActor::CommandFactory();
	my $command = $factory->create(@tokens);
	
	return $command;
}

# Keep Perl happy.
1;


lib/PerlActor/Command.pm
package PerlActor::Command;
use strict;
use base 'PerlActor::Object';
use fields qw( params );
use Error;
use PerlActor::Exception::AssertionFailure;
use PerlActor::Exception::NotImplemented;
use Carp;

#===============================================================================================
# Public Methods
#===============================================================================================

sub new
{
	my $proto = shift;
	my @params = @_;
	my $self = $proto->SUPER::new(@_);
	$self->{params} = \@params;
	return $self;
}

sub getParam
{
	my $self = shift;
	my $paramNumber = shift;
	return $self->{params}->[$paramNumber];
}

sub getParams
{
	my $self = shift;
	return @{$self->{params}};
}

sub execute
{
	my $self = shift;
	my $class = ref $self || $self;
	throw PerlActor::Exception::NotImplemented("command class '$class' does not implement execute()");
}

sub executeScript
{
	my $self = shift;
	my $file = shift;
	
	open SCRIPT, "$file"
		or throw PerlActor::Exception("cannot open test script file '$file': $!\n");
			
	my @lines = <SCRIPT>;
	my $script = new PerlActor::Script("$file");
	$script->setListener($self->{listener});
	$script->setLines(@lines);
	$script->execute();			
	close SCRIPT;
}

sub assert
{
	my $self = shift;
	my $condition = shift;
	my $message = shift;
	throw PerlActor::Exception::AssertionFailure($message)
		unless $condition;
}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

# Keep Perl happy.
1;


lib/PerlActor/Command/RunDir.pm
package PerlActor::Command::RunDir;
use strict;

use base 'PerlActor::Command';

use PerlActor::Exception;
use File::Find;

my @TESTS;

#===============================================================================================
# Public Methods
#===============================================================================================

sub execute
{
	my $self = shift;
	my $dir = $self->getParam(0);
	
	throw PerlActor::Exception("cannot collect test scripts: directory '$dir' does not exist!")
		unless -e $dir;
		
	$self->buildTestList($dir);

	foreach my $test (@TESTS)
	{
		$self->executeScript($test);
	}
}

sub buildTestList
{	
	my ($self, $dir) = @_;	
	find({ wanted => sub {$self->processFile}, follow => 1 }, $dir);
}

sub processFile
{	
	my $self = shift;	
	my $file = $File::Find::name;

	return unless $self->fileIsAPerlActorTest($file);
	
	push @TESTS, $file;
	
}

sub fileIsAPerlActorTest
{
	my ($self, $file) = @_;
	return $file =~ m/\.pact$/;
}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

# Keep Perl happy.
1;


lib/PerlActor/CommandFactory.pm
package PerlActor::CommandFactory;
use strict;
use base 'PerlActor::Object';

#===============================================================================================
# Public Methods
#===============================================================================================

sub create
{
	my $self = shift;
	my $commandName = shift;

	return $self->create('Null')
		unless $commandName;

	my @commandArgs = @_;
	my $error = '';
	foreach my $class ("PerlActor::Command::$commandName", $commandName)
	{
		unless ($error = $self->_compile($class))
		{
			return $class->new(@commandArgs) 
		}
		last unless ($error =~ /Can't locate/);
	}
	return $self->create('Unknown', $error);

}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

sub _compile
{
	my $self = shift;
	my $class = shift;
	return if (eval "require $class");
	return $@;
}

# Keep Perl happy.
1;


lib/PerlActor/Runner/Console.pm
package PerlActor::Runner::Console;
use strict;
use base 'PerlActor::Runner';
use fields qw( run passed failed aborted failedReport abortedReport started ended );

$| = 1;

use Benchmark qw( :hireswallclock );

#===============================================================================================
# Public Methods
#===============================================================================================

sub start
{
	my $self = shift;

	$self->_resetCounters();
	$self->_resetReports();
	$self->_printHeader();	
	$self->_startClock();
}

sub end
{
	my $self = shift;

	$self->_stopClock();

	my $runTime = timediff($self->{ended}, $self->{started});
	
	print "\n" . timestr($runTime);
	print "\nRun: $self->{run}, Passed: $self->{passed}, Failed: $self->{failed}, Aborted: $self->{aborted}.\n";
	
	$self->_printReports($self->{failedReport},'FAILED');
	$self->_printReports($self->{abortedReport},'ABORTED');

	unless ($self->{failed} or $self->{aborted})
	{
		print "\nOK ($self->{passed} scripts)";
	}
	print "\n";
}

sub _printReports
{
	my $self = shift;
	my $reports = shift;
	my $label = shift;
	
	if (@{$reports})
	{
		print "\n!!!$label!!!\n";
		my $count = 0;
		foreach my $entry (@{$reports})
		{
			$count++;
			print "\n$count) $entry\n";
		}
	}
}

sub scriptStarted
{ 
	my $self = shift;
	$self->{run}++;
}

sub scriptPassed
{ 
	my $self = shift;
	$self->{passed}++;
	print ".";
}

sub scriptAborted
{
	my ($self, $script, $exception) = @_;
	$self->{aborted}++;
	print "E";
	$self->_addReport($script, $exception, 'ABORTED', $self->{abortedReport});
}

sub scriptFailed
{ 
	my ($self, $script, $exception) = @_;
	$self->{failed}++;
	print "F";
	$self->_addReport($script, $exception, 'FAILED', $self->{failedReport});
}

#===============================================================================================
# Protected Methods - Don't even think about calling these from outside the class.
#===============================================================================================

sub _resetCounters
{
	my $self = shift;
	$self->{run}     = 0;
	$self->{passed}  = 0;
	$self->{failed}  = 0;
	$self->{aborted} = 0;
}

sub _resetReports
{
	my $self = shift;
	$self->{failedReport}  = [];
	$self->{abortedReport}  = [];
}

sub _addReport
{
	my ($self, $script, $exception, $label, $reportType) = @_;
	chomp $exception;
	push @{$reportType}, "$label: '$exception' " . $script->getTraceInfo();
}

sub _startClock
{
	my $self = shift;
	$self->{started} = new Benchmark();
}

sub _stopClock
{
	my $self = shift;
	$self->{ended} = new Benchmark();
}

sub _printHeader
{
	print "\nRunning Acceptance Tests at " . `date`;
	print "==========================================================\n";
}

# Keep Perl happy.
1;


lib/PerlActor.pm
package PerlActor;

use 5.008006;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '0.01';


# Preloaded methods go here.

1;
__END__

=head1 NAME

PerlActor - A simple automated executable acceptance test framework for Perl

=head1 DESCRIPTION

PerlActor is a simple automated executable acceptance test framework
for Perl.  It can be used by Extreme Programming (XP) teams to rapidly
develop executable acceptance tests for their Perl code.

XP mandates that the Customer writes acceptance tests for the application
under development.  The acceptance tests, also known as customer tests,
demonstrate that application features are complete and that they work as
expected.  Ideally, the tests should be directly executable and
automated so that they can be run continuously, without manual effort.

PerlActor enables an XP Customer to write *executable* acceptance
tests as plain text files, using a very simple syntax. Each test consists
of a number of parameterized commands (one per line), which PerlActor
uses to exercise the application. PerlActor parses the test, invokes the
commands with any parameters, and reports the result. The developers provide
glue code to implement the commands required by the tests.

PerlActor allows tests to be grouped into suites so that related tests can be
run as a group.  The Customer can also just place test scripts in a directory
structure and have PerlActor find and execute them all.

The PerlActor approach to acceptance testing has a number of advantages:

1) As the application grows the team will gradually develop a
comprehensive set of Commands for testing the application.

2) The Customer, independent of the development team, can write and execute
new tests for the application at any time, using any of the existing 
Commands.

3) Existing tests may be changed by the Customer at any time, again
without developer help.


=head1 EXAMPLE SCRIPT

The following sample test script is adapted from one of the example
scripts contained in the 'examples/calculator' directory of this
distribution.  It tests a toy calculator "application":

    # Script to check addition
    # 10 + 32 = 42

    # Create a new calculator application
    NewCalculator

    # Ensure that the display reads '0' intially
    CheckDisplayReads  0

    # Press key '1', then key '0'
    PressKeys          1  0
    CheckDisplayReads  10

    PressKeys          +
    CheckDisplayReads  10

    PressKeys          3  2
    CheckDisplayReads  32

    PressKeys          =
    CheckDisplayReads  42


The following is sample output from the included test runner:

    Running Acceptance Tests at Sun Apr 10 17:59:11 GMT 2005
    ==========================================================
    .....
    0.545077 wallclock secs ( 0.42 usr +  0.11 sys =  0.53 CPU)
    Run: 5, Passed: 5, Failed: 0, Aborted: 0.

and with test failure:

    Running Acceptance Tests at Sun Apr 10 18:04:36 GMT 2005
    ==========================================================
    .F...
    1.24811 wallclock secs ( 0.47 usr +  0.07 sys =  0.54 CPU)
    Run: 5, Passed: 4, Failed: 1, Aborted: 0.

    !!!FAILED!!!

    1) FAILED: 'Display is wrong: expected 41, got 42' in scripts/test_addition.pact
    at 'CheckDisplayReads  41', line 37


=head1 CREDITS

PerlActor is inspired by a similar open source framework for Java called 
'exactor' published by eXoftware [see http://exactor.sourceforge.net/].

=head1 SEE ALSO

PerlActor is very basic at present, and as you will no doubt have noticed,
the documentation is somewhat lacking.  This will change very soon.

In the meantime the PerlActor distribution comes with a simple example
showing how the framework is used for acceptance testing.  You'll find
it in the 'examples' directory of the distribution.

=head1 AUTHOR

Tony Byrne, E<lt>perl@byrnehq.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2005 by Tony Byrne

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.


=cut
